File: packages/android-compose/src/androidMain/kotlin/com/nostratech/modula/renderer/compose/SDUIRenderer.kt
package com.nostratech.modula.renderer.compose

import android.graphics.Color.parseColor
import com.nostratech.modula.*
import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.foundation.shape.RoundedCornerShape

@Composable
fun SDUIRenderer(components: List<SDUIComponent>) {
    Column {
        components.forEach { component ->
            SDUIComponentDispatcher(component)
        }
    }
}

@Composable
fun SDUIComponentDispatcher(component: SDUIComponent) {
    when (component) {
        is TextComponent -> TextRenderer(component)
        is ButtonComponent -> ButtonRenderer(component)
        is ContainerComponent -> ContainerRenderer(component)
        is ColumnComponent -> ColumnRenderer(component)
        is RowComponent -> RowRenderer(component)
    }
}

@Composable
fun TextRenderer(component: TextComponent) {
    val textAlign = when (component.style?.align) {
        "center" -> TextAlign.Center
        "end", "right" -> TextAlign.End
        "justify" -> TextAlign.Justify
        else -> TextAlign.Start
    }

    Text(
        text = component.text,
        fontSize = component.fontSize?.sp ?: 16.sp,
        color = component.color?.let { Color(parseColor(it)) } ?: Color.Unspecified,
        textAlign = textAlign,
        modifier = Modifier.fromStyle(component.style)
    )
}

@Composable
fun ButtonRenderer(component: ButtonComponent) {
    Button(
        onClick = { /* Handle action: component.action */ },
        contentPadding = PaddingValues(
            start = component.style?.padding?.left?.dp ?: 24.dp,
            top = component.style?.padding?.top?.dp ?: 12.dp,
            end = component.style?.padding?.right?.dp ?: 24.dp,
            bottom = component.style?.padding?.bottom?.dp ?: 12.dp
        ),
        shape = RoundedCornerShape(component.style?.round?.toIntOrNull()?.dp ?: 8.dp),
        colors = androidx.compose.material3.ButtonDefaults.buttonColors(
            containerColor = component.style?.backgroundColor?.let { Color(parseColor(it)) } ?: Color(parseColor("#007AFF")),
            contentColor = Color.White
        ),
        modifier = Modifier.padding(
            start = component.style?.margin?.left?.dp ?: 0.dp,
            top = component.style?.margin?.top?.dp ?: 0.dp,
            end = component.style?.margin?.right?.dp ?: 0.dp,
            bottom = component.style?.margin?.bottom?.dp ?: 0.dp
        )
    ) {
        Text(
            text = component.label, 
            fontSize = 16.sp, 
            fontWeight = androidx.compose.ui.text.font.FontWeight.SemiBold
        )
    }
}

@Composable
fun ContainerRenderer(component: ContainerComponent) {
    Box(modifier = Modifier.fromStyle(component.style)) {
        Column {
            component.children.forEach { child ->
                SDUIComponentDispatcher(child)
            }
        }
    }
}

@Composable
fun ColumnRenderer(component: ColumnComponent) {
    val style = component.style
    val verticalArrangement = when (style?.arrangement) {
        "center" -> Arrangement.Center
        "end", "bottom" -> Arrangement.Bottom
        "space-between" -> Arrangement.SpaceBetween
        "space-around" -> Arrangement.SpaceAround
        "space-evenly" -> Arrangement.SpaceEvenly
        else -> Arrangement.Top
    }
    val horizontalAlignment = when (style?.align) {
        "center" -> Alignment.CenterHorizontally
        "end" -> Alignment.End
        else -> Alignment.Start
    }

    Column(
        modifier = Modifier
            .fromStyle(style)
            .then(
                if (style?.scrollable == true)
                    Modifier.verticalScroll(rememberScrollState())
                else
                    Modifier
            ),
        verticalArrangement = verticalArrangement,
        horizontalAlignment = horizontalAlignment
    ) {
        component.children.forEach { child ->
            SDUIComponentDispatcher(child)
        }
    }
}

@Composable
fun RowRenderer(component: RowComponent) {
    val style = component.style
    val horizontalArrangement = when (style?.arrangement) {
        "center" -> Arrangement.Center
        "end", "right" -> Arrangement.End
        "space-between" -> Arrangement.SpaceBetween
        "space-around" -> Arrangement.SpaceAround
        "space-evenly" -> Arrangement.SpaceEvenly
        else -> Arrangement.Start
    }
    val verticalAlignment = when (style?.align) {
        "center" -> Alignment.CenterVertically
        "bottom" -> Alignment.Bottom
        else -> Alignment.Top
    }

    Row(
        modifier = Modifier
            .fromStyle(style)
            .then(
                if (style?.scrollable == true)
                    Modifier.horizontalScroll(rememberScrollState())
                else
                    Modifier
            )
        ,
        horizontalArrangement = horizontalArrangement,
        verticalAlignment = verticalAlignment
    ) {
        component.children.forEach { child ->
            SDUIComponentDispatcher(child)
        }
    }
}

@Composable
private fun Modifier.fromStyle(style: SDUIStyle?): Modifier {
    if (style == null) return this
    var modifier: Modifier = this

    // Margin (outer padding)
    style.margin?.let { m ->
        modifier = modifier.padding(
            start = m.left?.dp ?: 0.dp,
            top = m.top?.dp ?: 0.dp,
            end = m.right?.dp ?: 0.dp,
            bottom = m.bottom?.dp ?: 0.dp
        )
    }

    // Width
    val w = style.width
    modifier = when (w) {
        "match_parent", "fill" -> modifier.fillMaxWidth()
        "wrap_content" -> modifier.wrapContentWidth()
        null -> modifier
        else -> w.toIntOrNull()?.let { modifier.width(it.dp) } ?: modifier
    }

    // Height
    val h = style.height
    modifier = when (h) {
        "match_parent", "fill" -> modifier.fillMaxHeight()
        "wrap_content" -> modifier.wrapContentHeight()
        null -> modifier
        else -> h.toIntOrNull()?.let { modifier.height(it.dp) } ?: modifier
    }

    // Round / Radius (clip background)
    style.round?.let { r ->
        val radius = r.toIntOrNull()?.dp ?: 0.dp
        modifier = modifier.clip(RoundedCornerShape(radius))
    }

    // Background
    style.backgroundColor?.let { bgColor ->
        try {
            modifier = modifier.background(Color(parseColor(bgColor)))
        } catch (e: Exception) {
            // Invalid color
        }
    }

    // Padding (inner padding)
    style.padding?.let { p ->
        modifier = modifier.padding(
            start = p.left?.dp ?: 0.dp,
            top = p.top?.dp ?: 0.dp,
            end = p.right?.dp ?: 0.dp,
            bottom = p.bottom?.dp ?: 0.dp
        )
    }

    return modifier
}

File: packages/ios-swiftui/Sources/SDUI/SDUIRenderer.swift
import SwiftUI
import Shared // Importing the KMM shared framework

public struct SDUIRenderer: View {
    public let components: [SDUIComponent]
    
    public init(components: [SDUIComponent]) {
        self.components = components
    }
    
    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            ForEach(0..<components.count, id: \.self) { index in
                SDUIComponentDispatcher(component: components[index])
            }
        }
        .frame(maxWidth: .infinity, alignment: .topLeading)
    }
}

public struct SDUIComponentDispatcher: View {
    public let component: SDUIComponent
    
    public init(component: SDUIComponent) {
        self.component = component
    }
    
    public var body: some View {
        Group {
            if let textComp = component as? TextComponent {
                TextRenderer(component: textComp)
            } else if let buttonComp = component as? ButtonComponent {
                ButtonRenderer(component: buttonComp)
            } else if let containerComp = component as? ContainerComponent {
                ContainerRenderer(component: containerComp)
            } else if let columnComp = component as? ColumnComponent {
                ColumnRenderer(component: columnComp)
            } else if let rowComp = component as? RowComponent {
                RowRenderer(component: rowComp)
            } else {
                EmptyView()
            }
        }
    }
}

public struct TextRenderer: View {
    public let component: TextComponent
    
    public var body: some View {
        let alignStr = component.style?.align
        let textAlignment: TextAlignment = alignStr == "center" ? .center : (alignStr == "end" || alignStr == "right" ? .trailing : .leading)
        let frameAlignment: Alignment = alignStr == "center" ? .center : (alignStr == "end" || alignStr == "right" ? .trailing : .leading)

        Text(component.text)
            .font(.system(size: CGFloat(component.fontSize?.doubleValue ?? 16)))
            .foregroundColor(Color(hex: component.color ?? "#000000"))
            .multilineTextAlignment(textAlignment)
            // Expand to full width when align is set so alignment is visually visible
            .frame(maxWidth: alignStr != nil ? .infinity : nil, alignment: frameAlignment)
            .applySDUIStyle(component.style)
    }
}

public struct ButtonRenderer: View {
    public let component: ButtonComponent
    
    public var body: some View {
        Button(action: {
            print("Action: \(component.action)")
        }) {
            Text(component.label)
                .font(.system(size: 16, weight: .semibold))
                .padding(EdgeInsets(
                    top: CGFloat(component.style?.padding?.top?.doubleValue ?? 12),
                    leading: CGFloat(component.style?.padding?.left?.doubleValue ?? 24),
                    bottom: CGFloat(component.style?.padding?.bottom?.doubleValue ?? 12),
                    trailing: CGFloat(component.style?.padding?.right?.doubleValue ?? 24)
                ))
                .background(Color(hex: component.style?.backgroundColor ?? "#007AFF"))
                .foregroundColor(.white)
                .cornerRadius(CGFloat(Double(component.style?.round ?? "8") ?? 8))
        }
        .padding(EdgeInsets(
            top: CGFloat(component.style?.margin?.top?.doubleValue ?? 0),
            leading: CGFloat(component.style?.margin?.left?.doubleValue ?? 0),
            bottom: CGFloat(component.style?.margin?.bottom?.doubleValue ?? 0),
            trailing: CGFloat(component.style?.margin?.right?.doubleValue ?? 0)
        ))
    }
}

public struct ContainerRenderer: View {
    public let component: ContainerComponent
    
    public var body: some View {
        VStack(alignment: .leading) {
            SDUIRenderer(components: component.children)
        }
        .applySDUIStyle(component.style)
    }
}

public struct ColumnRenderer: View {
    public let component: ColumnComponent
    
    public var body: some View {
        let alignStr = component.style?.align
        let alignment = alignStr == "center" ? HorizontalAlignment.center : 
                        (alignStr == "end" || alignStr == "right" ? HorizontalAlignment.trailing : .leading)
        
        let arrangement = component.style?.arrangement
        let isSpaceBetween = arrangement == "space-between"
        let isSpaceAround = arrangement == "space-around" || arrangement == "space-evenly"
        let hasSpacers = isSpaceBetween || isSpaceAround

        // Only add surrounding spacers when explicitly requested (not default)
        let spacerTop = isSpaceAround || arrangement == "center" || arrangement == "end"
        let spacerBottom = isSpaceAround || arrangement == "center" || arrangement == "start"

        let content = VStack(alignment: alignment, spacing: hasSpacers ? nil : 10) {
            if spacerTop { Spacer(minLength: 0) }
            ForEach(0..<component.children.count, id: \.self) { index in
                SDUIComponentDispatcher(component: component.children[index])
                if index < component.children.count - 1 && hasSpacers {
                    Spacer(minLength: 0)
                }
            }
            if spacerBottom { Spacer(minLength: 0) }
        }
        
        if component.style?.scrollable == true {
            ScrollView { content }.applySDUIStyle(component.style)
        } else {
            content.applySDUIStyle(component.style)
        }
    }
}

public struct RowRenderer: View {
    public let component: RowComponent
    
    public var body: some View {
        let alignStr = component.style?.align
        let alignment = alignStr == "center" ? VerticalAlignment.center : 
                        (alignStr == "bottom" ? VerticalAlignment.bottom : .top)
        
        let arrangement = component.style?.arrangement
        let isSpaceBetween = arrangement == "space-between"
        let isSpaceAround = arrangement == "space-around" || arrangement == "space-evenly"
        let isCenterOrEnd = arrangement == "center" || arrangement == "end" || arrangement == "right"
        let hasSpacers = isSpaceBetween || isSpaceAround

        // For space-between/around: need full width, Spacers handle distribution
        // For center/end: leading Spacer, trailing Spacer respectively
        let spacerLeading = isSpaceAround || isCenterOrEnd
        let spacerTrailing = isSpaceAround || (arrangement == "center") || arrangement == nil || arrangement == "start"

        let content = HStack(alignment: alignment, spacing: hasSpacers ? nil : 10) {
            if spacerLeading { Spacer(minLength: 0) }
            ForEach(0..<component.children.count, id: \.self) { index in
                SDUIComponentDispatcher(component: component.children[index])
                if index < component.children.count - 1 && hasSpacers {
                    Spacer(minLength: 0)
                }
            }
            if spacerTrailing { Spacer(minLength: 0) }
        }
        // Spacers inside HStack only work if the HStack itself has full width
        .frame(maxWidth: hasSpacers || isCenterOrEnd ? .infinity : nil)
        
        if component.style?.scrollable == true {
            ScrollView(.horizontal) { content }.applySDUIStyle(component.style)
        } else {
            content.applySDUIStyle(component.style)
        }
    }
}

extension View {
    func applySDUIStyle(_ style: SDUIStyle?, overrideAlignment: Alignment? = nil) -> some View {
        guard let style = style else { return AnyView(self) }
        
        var view = AnyView(self)
        
        // 1. Padding (Inner spacing)
        if let p = style.padding {
            let edges = EdgeInsets(
                top: CGFloat(p.top?.doubleValue ?? 0),
                leading: CGFloat(p.left?.doubleValue ?? 0),
                bottom: CGFloat(p.bottom?.doubleValue ?? 0),
                trailing: CGFloat(p.right?.doubleValue ?? 0)
            )
            view = AnyView(view.padding(edges))
        }

        let alignStr = style.align
        let defaultAlignment: Alignment = alignStr == "center" ? .center : (alignStr == "end" || alignStr == "right" ? .trailing : .leading)
        let frameAlignment = overrideAlignment ?? defaultAlignment
        
        // 2. Dimensions (Frame) AFTER padding limits the background bounds
        let widthStr = style.width
        let heightStr = style.height
        let width = Double(widthStr ?? "")
        let height = Double(heightStr ?? "")
        
        if widthStr == "match_parent" || widthStr == "fill" {
            view = AnyView(view.frame(maxWidth: .infinity, alignment: frameAlignment))
        } else if let w = width {
            view = AnyView(view.frame(width: CGFloat(w), alignment: frameAlignment))
        }
        
        if heightStr == "match_parent" || heightStr == "fill" {
             view = AnyView(view.frame(maxHeight: .infinity, alignment: frameAlignment))
        } else if let h = height {
            view = AnyView(view.frame(height: CGFloat(h), alignment: frameAlignment))
        }

        // 3. Background (Fills the padded frame bounds)
        if let bgColor = style.backgroundColor {
            view = AnyView(view.background(Color(hex: bgColor)))
        }

        // 4. Corner Radius (Round) - Must be after background to clip it
        if let roundStr = style.round, let r = Double(roundStr) {
            view = AnyView(view.cornerRadius(CGFloat(r)))
        }

        // 5. Margin (Outer spacing outside of background clip)
        if let m = style.margin {
            let edges = EdgeInsets(
                top: CGFloat(m.top?.doubleValue ?? 0),
                leading: CGFloat(m.left?.doubleValue ?? 0),
                bottom: CGFloat(m.bottom?.doubleValue ?? 0),
                trailing: CGFloat(m.right?.doubleValue ?? 0)
            )
            view = AnyView(view.padding(edges))
        }
        
        return view
    }
}

// Helper for Color from Hex
extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (1, 1, 1, 0)
        }

        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue:  Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
}
